service/ethevents.go

```diff
@@ -40,20 +41,37 @@ func EthEvents(
 		specs.Contracts,
 		specs.NetworkSource,
 		signer,
-		w3uri,
+		w3uris,
 		cm,
 		vocapp,
 		ethereumWhiteList,
 	)
+
 	if err != nil {
 		return fmt.Errorf("couldn't create ethereum events listener: %w", err)
 	}
 	for _, e := range evh {
 		ev.AddEventHandler(e)
 	}
-	go func() {
-		ev.SubscribeEthereumEventLogs(ctx, startBlock)
-	}()
+
+	initialized := new(bool)
+	go func(initialized *bool) {
+		for {
+			if ctx.Err() != nil {
+				return
+			}
+			ctx, cancel := context.WithCancel(ctx)
+			if *initialized {
+				if ev.VotingHandle, err = ethereumhandler.NewEthereumHandler(ev.ContractsInfo, specs.NetworkSource, w3uris); err != nil {
+					log.Fatalf("cannot restart ethereum events, cannot create ethereum handler: %v", err)
+				}
+			}
+			*initialized = true
```

Comment:
I don't understand the logic behind this routine. Why a for loop? "initialized" is set to true on each iteration, why? Also why ev.VotingHandle is inside the for loop?

Reply:
kind of a workaround because new EthEvents already invokes new ethereum handler and I wanted to avoid this. Solving properly requires some refactor on the new EthEvents function. This is out of the scope of this commit.
WIll add a TODO explaining more about this for adding clarity and i'll make another PR improving this.

Reply:
will be better addressed on the next pr